### ISSUE TITLE
Multi Schedulers don't synchronize task instance state when using Triggerer 

### ISSUE THREAD

**Issue by Greetlist** [apache/airflow#20308]
### Apache Airflow version

2.2.1

### What happened

I find some Deferrable Operators are marked failed unexpectedly. When I clear these tasks, some of them will success, but some of them will fail as before.Even I clear the same task, sometime it is marked success, sometime it is marked failed.
I find this log in scheduler log
```
ERROR[0m - Executor reports task instance <TaskInstance: x [queued]> finished (success) although the task says its queued. (Info: None) Was the task killed externally?[0m
[[34m2021-12-14 10:22:20,208[0m] {[34mtaskinstance.py:[0m1706} ERROR[0m - Executor reports task instance <TaskInstance: x [queued]> finished (success) although the task says its queued. (Info: None) Was the task killed externally?[0m
[[34m2021-12-14 10:22:20,215[0m] {[34mtaskinstance.py:[0m1281} INFO[0m - Marking task as FAILED. dag_id=x, task_id=x, execution_date=20211212T170000, start_date=20211214T102218, end_date=20211214T102220[0m
```

### What you expected to happen

When I clear a Deferrable Operator task, task final state must be the same instead of sometime it is marked success, sometime it is marked failed.

### How to reproduce

Create a Deferrable Operator and this operator's run function **will yield immediately**.Like this:
```
class ExampleTrigger(BaseTrigger):
    async def run(self):
        while datetime.now() > datetime.strptime("1900-01-01", "%Y-%m-%d"):
            await asyncio.sleep(1)
        yield TriggerEvent('success')
```
In my case, I just query a task's state from another airflow database, and the execution_date(old version airflow) is prev_day [*would yield immediately*]

This task will be mark failed by scheduler randomly.

### Operating System

Ubuntu 16.04

### Deployment

Other

### Deployment details

1. One Triggerer process.
2. Two Schedulers.
3. One WebServer.

### Anything else

I check the code of:
1. scheduler
2. trigger
3. taskinstance
4. timestamp of log

and figure out what happend in this condition:

Assume TA is a deferrable task.
1. Scheduler loop 1 change TA task state: SCHEDULED -> QUEUED -> RUNNING -> DEFERRED, but this scheduler loop do not quit.
![image](https://user-images.githubusercontent.com/18439743/146117373-fb3a11cb-4e17-4b17-b621-f81dd33d0100.png)
2. Meanwhile Trigger change TA task state: DEFERRED -> SCHEDULED
3. Scheduler loop 2 start, find the TA task is SCHEDULED and then change TA task state: SCHEDULED -> QUEUED
4. Scheduler loop 1 start to run the *_process_executor_events* function to check Worker-Process return status, and fetch a **WRONG state QUEUED** from database, finally run into this code, and mark the task as failed.
![image](https://user-images.githubusercontent.com/18439743/146113442-c76de06c-887c-45bc-b289-6110141efd7a.png)
5. Scheduler loop 2 enqueue the *airflow task run xxx* command and then worker also fetch a failed task to execute.
![image](https://user-images.githubusercontent.com/18439743/146114776-57041117-0593-408b-bb9d-f6182deb0562.png)

In my opinion, I think there are two reasons cause this problem:
1. the query in *_executable_task_instances_to_queued* function:
![image](https://user-images.githubusercontent.com/18439743/146115401-d5655678-88f2-4050-adec-9dd14f6262b7.png)
can not guarantee one task is executed only in one scheduler loop.
2. The Trigger-Process can modify this task state to SCHEDULED asynchronously.

**How to solve this problem**
1. Add a new column in task_instance table named is_still_processing default is False
2. While query SCHEDULED task, add a filter like ```TI.is_still_processing == False``` at
![image](https://user-images.githubusercontent.com/18439743/146120873-9aba3ce5-b669-40a5-b9dc-3b7148247b84.png)
3. Before enqueue task instance to executor queue, change is_still_processing to True in database.
4. At the end of *_executable_task_instances_to_queued* loop:
![image](https://user-images.githubusercontent.com/18439743/146120181-4e3d92a4-2767-48b7-a2e2-2a51f0fb7acc.png)
change this task's is_still_processing to False in database.

If I can guarantee one Deferrable Operator task is executed only in one scheduler loop and quit successfully,I can solve this problem.
I think only Deferrable Operator task which can yield immediately have this problem.

or 
Just change task instance state to SUCCESS instead of SCHEDULED, but this Deferrable task's next_method has no chance to be executed.

Any Suggestions? I need to solve this problem ASAP.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by Greetlist** [comment]
At this moment, I solve this problem by this:

![image](https://user-images.githubusercontent.com/18439743/148174419-5f27d495-ef6c-4a08-b262-c41ada4725ad.png)

Pass *custom_skip* args to method_name when invoking the **defer** function.
![image](https://user-images.githubusercontent.com/18439743/148174785-3d248b35-f9e6-4e5b-88bd-3328553a2eea.png)

Tricky solution.

**Comment by akaa07** [comment]
I am having the same problem.
Multiple delayable operators randomly terminate abnormally.
Since I am using MWAA, there is no workaround solution.

**Comment by potiuk** [comment]
cc: @dstandish @andrewgodwin - seems like a likely case to take a look at.

**Comment by andrewgodwin** [comment]
Hmm, I would have thought the scheduler row locking would have prevented this. What database are you using to run against?

**Comment by Greetlist** [comment]
The DataBase is in K8S, and Image is: ```Percona Server 8.0.17```
![image](https://user-images.githubusercontent.com/18439743/156089137-9cbb7dff-ccdd-498a-a115-6e64a3ee3563.png)

I make a test on test DataBase whose version is ```(MySQL Community Server - GPL 8.0.26)```.
![image](https://user-images.githubusercontent.com/18439743/156101859-cb010683-93f6-4c7a-96d4-ce3b45ee5324.png)

``` for update ``` do lock these rows when I try to modify one of these rows at another session. 
As I comment before, I fix it with a tricky way.Prevent the deferrable task is executed twice, this problem is never show again.

It seem that the trigger process do modify task_instance record even the record locked by ``` begin; for update; commit ```

**Comment by potiuk** [comment]
@andrewgodwin - maybe there is an extra commit somewhere that releases the lock. I know @ashb  implemented mechanisms to discover that that has been used in scheduler row-level locking - maybe we should use it here (unless it is done already ?)

**Comment by ashb** [comment]
https://github.com/apache/airflow/blob/aa2cb5545f09d694b9143b323efcd4f6b6c66e60/airflow/utils/sqlalchemy.py#L284-L303 if it is useful.

**Comment by Greetlist** [comment]
I doubt if this code would release the lock.
![image](https://user-images.githubusercontent.com/18439743/156513818-4e2eb2ed-bd5e-4f7b-b621-5a830d9458c6.png)

![image](https://user-images.githubusercontent.com/18439743/156513992-5307a02e-dd34-4610-ba86-b3a8033ad399.png)

**Comment by ashb** [comment]
The lock is not an object -- it is a database side thing tied to the open transaction.  The only way of releasing the lock is to COMMIT or ROLLBACK the DB session.

**Comment by potiuk** [comment]
> The lock is not an object -- it is a database side thing tied to the open transaction. The only way of releasing the lock is to COMMIT or ROLLBACK the DB session.

Yeah. It seems that even you can't explicitly release such lock even if you want, so it would be impossible to implement.

**Comment by ashb** [comment]
(I haven't been following the context, so this suggestion may make no sense).

One option might be to use savepoints -- 

```sql
SAVEPOINT before_lock;
SELECT * from trigger FOR UPDATE;
ROLLBACK TO before_lock;
-- Lock would be released now.
```

That rollback would only work if there have been no changes flushed.

(SQLA has the means to do this as "nested" transactions so we don't need to resort to raw SQL)

**Comment by Greetlist** [comment]
I re-scan the **scheduler_job.py, models/trigger.py ** again.
``` _do_scheduling ``` function has RowLock, ``` _process_executor_events ``` function also has RowLock and two functions use the same session.

https://github.com/apache/airflow/blob/17d3e78e1b4011267e81846b5d496769934a5bcc/airflow/models/trigger.py#L109 has no RowLock, it will select some records which already selected by scheduler.But it will update these records, will wait the RowLock, wait until one scheduler loop is done.

The process above is make sense. But it do has inconsistent problem.I have no idea where the key is.
Confusing.

**Comment by jskalasariya** [comment]
Any other work around apart from the one which was suggested by @Greetlist ?

I am also running into same problem when I ran two scheduler.

**Comment by Greetlist** [comment]
@jskalasariya If you also use the deferrable task, you can fix this via https://github.com/apache/airflow/issues/20308#issuecomment-1005435631 temporarily. You can just try to change these codes and make some tests to checkout whether this problem is disappear.
At this moment, In my product env:
**airflow version**:
> 2.2.0

**airflow components**:
> 1. four schedulers
> 2. three triggerers
> 3. two web servers

**operating system**:
> Ubuntu16.04

**machine resource**:
> 64G memory
> 6Core(2slots * 3Core) NUMA Disabled

**database**:
> percona-server: 8.0.17 in k8s cluster as Deployment

works as expected.
By the way, if you want to use the deferrable operator's defer_timeout parameter，you may have some **dead-lock** problems:
```models/trigger.py[func submit_* clean_unused]``` will hold row lock **AND** ```jobs/scheduler_job.py[func check_trigger_timeouts])``` will hold another row lock. I fix this dead-lock by adding ```with_row_locks```,  ```order by``` and split query and update into two steps:
![image](https://user-images.githubusercontent.com/18439743/167544836-b586eb46-9516-41ba-a627-df70e7fc98ae.png)

**Comment by masterlittle** [comment]
Is this planned to be fixed? This is a major issue and causing failures and operational nightmares

**Comment by jskalasariya** [comment]
@ashb @potiuk are we prioritising this issue?

**Comment by potiuk** [comment]
No idea @jskalasariya . There is nothing like prioritization in Airflow. Things get fixed when someone fix them because that someone decides it is important. If it is not fixed, then it means no-one though it is important enough to fix it. Not sure why you asking us 2 for prioritisation.

**Comment by JJJzheng** [comment]
Hi all
I met the same problem.
And my airflow version is airflow-2.2.5.

**Comment by alon050** [comment]
does this means, that when you have a triggerer dag you cant have more than 1 scheduler running?

**Comment by potiuk** [comment]
Hard to say. @andrewgodwin maybe it could be triaged better? sounds like a number of people get impacted.

**Comment by andrewgodwin** [comment]
Hmm, I intended this code to work across multiple schedulers when I wrote it but it's been long enough since I looked at it that I don't remember how I intended that! It's quite possible it needs more locking.

**Comment by hussein-awala** [comment]
I was looking into this issue and attempted to resolve it. However, I discovered that it has already been addressed in #23846, which was released in version 2.3.3. This version is newer than the reported versions (2.2.1 and 2.2.5).
I will close it for now, but we can reopen it if necessary.
